Merge template 2.2 and update to latest DSL2 syntax

JoseEspinosa opened the pull request:
# nf-core/chipseq pull request

* Merge branch `nf-core-template-merge-2.2`
* Update modules to the latest version
* Bring back the skip `params`
* Refactor modules.config

## PR checklist

- [x] This comment contains a description of changes (with reason)
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If necessary, also make a PR on the [nf-core/chipseq branch on the nf-core/test-datasets repo](https://github.com/nf-core/test-datasets/pull/new/nf-core/chipseq)
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Make sure your code lints (`nf-core lint .`).
- [ ] Documentation in `docs` is updated
- [x] `CHANGELOG.md` is updated
- [ ] `README.md` is updated

**Learn more about contributing:** [CONTRIBUTING.md](https://github.com/nf-core/chipseq/tree/master/.github/CONTRIBUTING.md)

drpatelh commented on .github/workflows/awsfulltest.yml:
```diff
@@ -26,5 +26,5 @@ jobs:
-          profiles: '[ "test_full", "aws_tower" ]'
-
+          profiles: test_full,aws_tower
+          pre_run_script: 'export NXF_VER=21.10.3'
```
```suggestion
```

drpatelh replied:
Should be able to remove this now because the latest version on Tower has been bumped.

drpatelh commented on .github/workflows/awstest.yml:
```diff
@@ -11,18 +11,18 @@ jobs:
-          profiles: '[ "test", "aws_tower" ]'
-
+          profiles: test,aws_tower
+          pre_run_script: 'export NXF_VER=21.10.3'
```
```suggestion
```

drpatelh commented on CHANGELOG.md:
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+    * Removed `--publish_dir_mode` as it is no longer required for the new syntax
+* Bump minimum Nextflow version from `21.04.0` -> `21.10.3`
+* Updated pipeline template to [nf-core/tools 2.2](https://github.com/nf-core/tools/releases/tag/2.2)
 * Pipeline has been re-implemented in [Nextflow DSL2](https://www.nextflow.io/docs/latest/dsl2.html)
 * Updated Nextflow version to `v21.04.0` (see [nextflow#572](https://github.com/nextflow-io/nextflow/issues/1964))
```
```suggestion
```

drpatelh commented on CHANGELOG.md:
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+    * Removed `--publish_dir_mode` as it is no longer required for the new syntax
+* Bump minimum Nextflow version from `21.04.0` -> `21.10.3`
+* Updated pipeline template to [nf-core/tools 2.2](https://github.com/nf-core/tools/releases/tag/2.2)
 * Pipeline has been re-implemented in [Nextflow DSL2](https://www.nextflow.io/docs/latest/dsl2.html)
 * Updated Nextflow version to `v21.04.0` (see [nextflow#572](https://github.com/nextflow-io/nextflow/issues/1964))
 * Move template boilerplate code to Groovy `lib/`
```
```suggestion
```

drpatelh replied:
Implied in the points above already

drpatelh commented on assets/samplesheet.csv:
```diff
@@ -0,0 +1,3 @@
+sample,fastq_1,fastq_2
```
Need to change the format to look like the one for this pipeline.

JoseEspinosa replied:
Actually there is an `assets/samplesheet_pe.csv` and a `assets/samplesheet_se.csv` files created by you, should we just remove this one?

drpatelh commented on modules/local/get_chrom_sizes.nf:
```diff
@@ -1,20 +1,10 @@
-
-params.options = [:]
-
 process GET_CHROM_SIZES {
```
Can delete and install this module from here https://github.com/nf-core/modules/blob/master/modules/custom/getchromsizes/main.nf

JoseEspinosa replied:
Addressed in 318dd3c298e582f87554bee9b33f99a92333a3a4

drpatelh commented on modules/local/igv.nf:
```diff
@@ -42,7 +32,7 @@ process IGV {
 
     cat <<-END_VERSIONS > versions.yml
-    ${getProcessName(task.process)}:
+    "${task.process}":
```
What about the commented out lines at the bottom of this script? https://github.com/JoseEspinosa/nf-core-chipseq/blob/865210afa074301fad8e83429dbae93b762ea0ae/modules/local/igv.nf#L40-L41

JoseEspinosa replied:
TBH, they were already there before I started reimplementing the pipeline and thought that were lines that at some point will be reused again

drpatelh commented on modules/local/macs2_consensus.nf:
```diff
@@ -34,7 +23,7 @@ process MACS2_CONSENSUS {
     script: // This script is bundled with the pipeline, in nf-core/chipseq/bin/
     if (meta.multiple_groups || meta.replicates_exist) {
-        def prefix       = options.suffix ? "${meta.id}${options.suffix}.consensus_peaks" : "${meta.id}.consensus_peaks"
+        def prefix       = task.ext.prefix ? "${task.ext.prefix}.consensus_peaks": "${meta.id}.consensus_peaks"
```
Maybe this should be set in `modules.config` instead? 
```suggestion
        def prefix = task.ext.prefix ?: "${meta.id}"
```

JoseEspinosa replied:
Addressed in 38769b6be5eab09c037e45f6292daaba60b4db3f

drpatelh commented on modules/local/multiqc.nf:
```diff
@@ -63,17 +52,18 @@ process MULTIQC {
     path "versions.yml"        , emit: versions
```
Need to remove this line? https://github.com/JoseEspinosa/nf-core-chipseq/blob/865210afa074301fad8e83429dbae93b762ea0ae/modules/local/multiqc.nf#L46

JoseEspinosa replied:
Done here 154674d9f939ee4195c0213ed2466ed996363d7f

drpatelh commented on modules/local/multiqc_custom_peaks.nf:
```diff
@@ -26,8 +15,7 @@ process MULTIQC_CUSTOM_PEAKS {
+    def prefix = task.ext.prefix ?: "${meta.id}"
     """
     cat $peak | wc -l | awk -v OFS='\t' '{ print "${prefix}", \$1 }' | cat $peak_count_header - > ${prefix}.peak_count_mqc.tsv
     cat $frip_score_header $frip > ${prefix}.FRiP_mqc.tsv
```
Output `versions.yml` here too?

JoseEspinosa replied:
Added here aa10f83e683a162c0b69cc43bc3e16b379512cb4

drpatelh commented on modules/local/multiqc_custom_phantompeakqualtools.nf:
```diff
@@ -1,21 +1,9 @@
-        mode: params.publish_dir_mode,
-        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:[:], publish_by_meta:[]) }
-
     //TODO substitute with a newest tag (see https://github.com/BioContainers/containers/issues/416)
```
Remove/address this comment?

JoseEspinosa replied:
Addressed here 50c0410850b88b2278128d65118a5c283022092b

drpatelh commented on modules/local/plot_homer_annotatepeaks.nf:
```diff
@@ -1,24 +1,13 @@
-params.options = [:]
-options        = initOptions(params.options)
-
 /*
  * Aggregated QC plots for peak-to-gene annotation
  */
```
```suggestion
```

drpatelh commented on modules/local/plot_macs2_qc.nf:
```diff
@@ -1,24 +1,13 @@
-params.options = [:]
-options        = initOptions(params.options)
-
 /*
  * Aggregated QC plots for peaks
  */
```
```suggestion
```

drpatelh commented on nextflow.config:
```diff
@@ -63,30 +66,27 @@ params {
-    igenomes_ignore            = false
     validate_params            = true
     show_hidden_params         = false
     schema_ignore_params       = 'genomes,modules'
```
```suggestion
    schema_ignore_params       = 'genomes'
```

drpatelh commented on nextflow.config:
```diff
@@ -205,10 +206,13 @@ manifest {
     homePage        = 'https://github.com/nf-core/chipseq'
```
Add our names in the author field above :)

JoseEspinosa replied:
Done in 9eb3c17951b83774e42889b4797ef28ae203945f :boom: 
BTW, your name is not in the manifest of the rnaseq pipeline :scream:

drpatelh reviewed (approved):
Fell short of looking at `modules.config` and `chipseq.nf` @JoseEspinosa. Will have to look at those another time. Otherwise really great job 🚀 🤩 

Be great if you are able to do some tests with the current release and the one on `dev` to see if the files publised are the same e.g. when setting all of the `--save_*` parameters and when running the pipeline with default options e.g. https://nfcore.slack.com/archives/CE8SSJV3N/p1639388965355400

And even better if you are able to compare the final outputs to see if they are the same/similar e.g. peaks, consensus peaks etc

JoseEspinosa commented:
Thanks @drpatelh ! I addressed most of your comments and left a couple of replies to them. I have been checking that all the `--skip_` params worked and no warmings were thrown but did not test so thoroughly t the `--save_* params. Will do and compare outputs. I will merge this PR and keep working on the merged version :candy:
